File: rust/2015/src/day17.rs
use itertools::Itertools;
use std::ops::{Add, Mul};

fn dot_product<T>(a: &[T], b: &[T]) -> T
where
    T: Copy + Default + Add<Output = T> + Mul<Output = T>,
{
    assert_eq!(a.len(), b.len());
    let mut acc = T::default();
    for (&x, &y) in a.iter().zip(b) {
        acc = acc + x * y;
    }
    acc
}

fn solve_case<const LITERS: u32, const COUNT_WAYS: bool>(filename: &str) -> u32 {
    let containers = std::fs::read_to_string(filename)
        .unwrap()
        .lines()
        .map(|line| line.parse().unwrap())
        .collect::<Vec<u32>>();
    let mut count = [0; 21];
    for combination in std::iter::repeat([0, 1])
        .take(containers.len())
        .multi_cartesian_product()
    {
        let exact_match = dot_product(&containers, &combination) == LITERS;
        if (!COUNT_WAYS) {
            count[0] += exact_match as u32;
        } else if (exact_match) {
            let num_used = combination.iter().sum::<u32>();
            count[num_used as usize] += 1;
        }
    }
    if (!COUNT_WAYS) {
        return count[0];
    } else {
        return *count.iter().skip_while(|c| **c == 0).next().unwrap();
    }
}

fn main() {
    println!("Part 1");
    assert_eq!(4, solve_case::<25, false>("day17.example"));
    assert_eq!(1304, solve_case::<150, false>("day17.input"));
    println!("Part 2");
    assert_eq!(3, solve_case::<25, true>("day17.example"));
    assert_eq!(18, solve_case::<150, true>("day17.input"));
}

File: 2015/day17.cpp
// https://adventofcode.com/2015/day/17

#include "../common/common.h"
#include "../common/rust.h"

#ifndef AOC_IMPORT_STD
#include <algorithm>
#include <print>
#include <ranges>
#endif

Vec<u32> parse(String const& filename) {
  return aoc::views::read_numbers<u32>(filename) | aoc::collect_vec<u32>();
}

template <u32 LITERS, bool COUNT_WAYS>
fn solve_case(Vec<u32> const& sizes) -> u32 {
  let containers = [&] {
    auto array = std::array<u32, (LITERS > 25) ? 20 : 5>{};
    stdr::copy(sizes, aoc::inserter_it(array));
    return array;
  }();
  let size = containers.size();
  auto count = std::array<u32, (COUNT_WAYS ? size : 0) + 1>{};
  for (let& group : containers | aoc::views::binary_combinations()) {
    let sum = aoc::ranges::accumulate(group | aoc::views::deref, 0u);
    if (sum != LITERS) {
      continue;
    }
    let num_used = static_cast<u32>(group.size());
    if constexpr (!COUNT_WAYS) {
      count[0] += 1;
    } else {
      count[num_used] += 1;
    }
  }
  if constexpr (!COUNT_WAYS) {
    return count[0];
  } else {
    return *stdr::find_if(count, aoc::not_equal_to_value<u32>{0});
  }
}

int main() {
  std::println("Part 1");
  let example = parse("day17.example");
  AOC_EXPECT_RESULT(4, (solve_case<25, false>(example)));
  let input = parse("day17.input");
  AOC_EXPECT_RESULT(1304, (solve_case<150, false>(input)));

  std::println("Part 2");
  AOC_EXPECT_RESULT(3, (solve_case<25, true>(example)));
  AOC_EXPECT_RESULT(18, (solve_case<150, true>(input)));

  AOC_RETURN_CHECK_RESULT();
}
